src/js/classes/BackgroundFetch.js

```diff
@@ -0,0 +1,84 @@
+import getURLsForDownload from '../utils/getURLsForDownload';
+
+const BG_FETCH_ID_TEMPLATE = /kino-(?<videoId>[a-z-_]+)/;
+
+export default class BackgroundFetch {
+  /**
+   * Construct a new Background Fetch wrapper.
+   */
+  constructor() {
+    this.id = '';
+    this.videoId = '';
+    this.onprogress = () => {};
+    this.ondone = () => {};
+  }
+
+  async maybeAbort(swReg) {
+    const existingBgFetch = await swReg.backgroundFetch.get(this.id);
+    if (existingBgFetch) await existingBgFetch.abort();
+  }
+
+  /**
+   *
+   * @param {BackgroundFetchRegistration} registration Background Fetch Registration object.
+   */
+  fromRegistration(registration) {
+    const matches = registration.id.match(BG_FETCH_ID_TEMPLATE);
+
+    this.id = registration.id;
+    this.videoId = matches.groups?.videoId || '';
+  }
+
+  async start(videoData) {
+    this.videoId = videoData.id;
+    this.id = `kino-${this.videoId}`;
+
+    const urls = await getURLsForDownload(this.videoId);
+
+    /** @type {Promise<Response[]>} */
+    const requests = urls.map((url) => fetch(url, { method: 'HEAD' }));
+
+    Promise.all(requests).then((responses) => {
+      const sizes = responses.map((response) => response.headers.get('Content-Length'));
+      const downloadTotal = sizes.includes(null)
+        ? null
+        : sizes.reduce((total, size) => total + parseInt(size, 10), 0);
+
+      // eslint-disable-next-line compat/compat
+      navigator.serviceWorker.ready.then(async (swReg) => {
+        this.maybeAbort(swReg);
```

Comment:
I guess you'd want to await there before or is this intentional?
```suggestion
        await this.maybeAbort(swReg);
```

Reply:
Added in 7979e32.